### ISSUE TITLE
about gt-ucf.npy

### ISSUE THREAD

**Issue by qzfm** [henrryzh1/UR-DMU#13]
Excuse me, why is the size of frame labels in gt-ucf.npy is 1109680 instead of 1109824 (if drop the last frames). I drop the last few frames of each video so that the total frame count for each video is divisible by 16, but the final total frame count is 1109824.

It seems that when the number of frames in a video is exactly divisible by 16, the corresponding .npy file is missing the last 16 frames.
For example, video "Assault/Assault006_x264" has 8096 frames, but the Assault006_x264.npy only has 505 clips. (505 * 16 = 8080).

**Comment by henrryzh1** [comment]
Thank you for your attention. After extracting the features, you can recreate the gt-ucf.npy  file.

**Comment by jqjq726** [comment]
hi @qzfm may I know are u able to recreate the gt-ucf.npy file at last? as I don't see a file of generating the ground truth files here

**Comment by qzfm** [comment]
> hi [@qzfm](https://github.com/qzfm) may I know are u able to recreate the gt-ucf.npy file at last? as I don't see a file of generating the ground truth files here

@jqjq726 
Hello, I finally wrote a python script myself to recreate a new gt-ucf.npy based on list/UCF_Annotation.txt and the 10-crop feature files I extracted.

**Comment by jqjq726** [comment]
@qzfm Would you be willing to share your Python script for generating the gt-ucf.npy file? It would be very helpful for me, thanks!

**Comment by qzfm** [comment]
> [@qzfm](https://github.com/qzfm) Would you be willing to share your Python script for generating the gt-ucf.npy file? It would be very helpful for me, thanks!


@jqjq726 Okay, I will share this Python script with you. However, since several months have passed, some files have been deleted, and I’m not sure if this script is the exact version I used to create gt-ucf.npy at the time. 
Hope this helps! Please note that the line `num_frames = num_frames - (num_frames % 16)` should be adjusted based on the actual video features you extracted. For example, you might discard the last few frames or duplicate the last frame—there could be different ways to handle this.

```
import os
import numpy as np
from tqdm import tqdm
ori_label = 'UR-DMU/list/UCF_Annotation.txt'
test_list = 'UR-DMU/list/ori_label/UCF_Test.list'
gt_path = 'UR-DMU/list/ori_label/UCF_GT.npy'
with open(ori_label, 'r') as f1:

    gt = np.array([], dtype=float)

    dict_annotation = {}

    lines = f1.readlines()
    for line in lines:
        line = line.strip()
        items = line.split(' ')

        # RoadAccidents/RoadAccidents021_x264.mp4 155 RoadAccidents 30 90 -1 -1 
        name = items[0].split('.mp4')[0] # RoadAccidents/RoadAccidents021_x264
        dict_annotation[name] = line
        

    with open(test_list, 'r') as f2:
        test_lines = f2.readlines()
        print(len(test_lines))

        i = 0
        for line in tqdm(test_lines):
            
            if i % 10 != 0:
                i += 1
                continue

            i += 1
                
            # dataset/Abuse/Abuse028_x264__0.npy
            items = line.strip().split('/')

            name = items[-2] + '/' + items[-1].split('__')[0]
            
            # # RoadAccidents/RoadAccidents021_x264.mp4 155 RoadAccidents 30 90 -1 -1
            infos = dict_annotation[name].strip().split(' ')
            num_frames = int(infos[1])
            event1 = (int(infos[3]), int(infos[4]))
            event2 = (int(infos[5]), int(infos[6]))

           
            num_frames = num_frames - (num_frames % 16) # 要根据实际特征文件进行修改


            if event1[0] != -1:
                event1 = (min(event1[0], num_frames), min(event1[1], num_frames))
            if event2[0] != -1:
                event2 = (min(event2[0], num_frames), min(event2[1], num_frames))

            # 创建一个全为0的数组
            event_array = np.zeros(num_frames, dtype=float)

            # 标记event1的时间段
            if event1[0] != -1:
                event1_start = min(event1[0], num_frames)
                event1_end = min(event1[1], num_frames)
                event_array[event1_start:event1_end] = 1.0

            # 标记event2的时间段
            if event2[0] != -1:
                event2_start = min(event2[0], num_frames)
                event2_end = min(event2[1], num_frames)
                event_array[event2_start:event2_end] = 1.0

            gt = np.concatenate([gt, event_array])

        print(gt.shape)
        np.save(gt_path, gt)
```

**Comment by jqjq726** [comment]
@qzfm Sure! Thanks for providing the code! I really appreciate it. I'll explore it, thanks!

**Comment by jqjq726** [comment]
@qzfm , can I ask do u extract the I3D features yourself? As I was using the I3d 10 crop features dataset link of UCF-Crime provided by the author (https://github.com/Roc-Ng/DeepMIL), but also i got the problem where my ground truth shape generated is (1109824,) but the ground truth they generated has the shape of (1109680,)

**Comment by qzfm** [comment]
> [@qzfm](https://github.com/qzfm) , can I ask do u extract the I3D features yourself? As I was using the I3d 10 crop features dataset link of UCF-Crime provided by the author (https://github.com/Roc-Ng/DeepMIL), but also i got the problem where my ground truth shape generated is (1109824,) but the ground truth they generated has the shape of (1109680,)


Yes, I encountered this problem a few months ago. In fact, I also tried using the I3D features from the (https://github.com/Roc-Ng/DeepMIL) repository and found that their features had inconsistent processing methods for some videos, as I mentioned in this issue. If you use the features from the (https://github.com/Roc-Ng/DeepMIL) repository, I remember that there is no need to  recreate the gt-ucf.npy file.

**Comment by jqjq726** [comment]
I see, as I was trying to implement multi classification model so that the model able to detect which types of anomalous events instead of just normal vs anomalous, i had to generate a multi class labeled ground truth, so i have to first know how to generate the binary labelled ground truth first. Can I ask if u have experience on developing the multi classification model?

> > [@qzfm](https://github.com/qzfm) , can I ask do u extract the I3D features yourself? As I was using the I3d 10 crop features dataset link of UCF-Crime provided by the author (https://github.com/Roc-Ng/DeepMIL), but also i got the problem where my ground truth shape generated is (1109824,) but the ground truth they generated has the shape of (1109680,)
> 
> Yes, I encountered this problem a few months ago. In fact, I also tried using the I3D features from the (https://github.com/Roc-Ng/DeepMIL) repository and found that their features had inconsistent processing methods for some videos, as I mentioned in this issue. If you use the features from the (https://github.com/Roc-Ng/DeepMIL) repository, I remember that there is no need to recreate the gt-ucf.npy file.

**Comment by qzfm** [comment]
@jqjq726 
You can modify the Python script and set the labels based on video names, for example, normal events as 0, Abuse as 1, Arrest as 2, and so on. I did try some multi-class classification tasks, but the results were not very good. I feel that this might be because UCF-Crime is not a high-quality dataset. Additionally, I remember that VadCLIP involves multi-class classification tasks, so you can take a look and learn how it is implemented.

**Comment by jqjq726** [comment]
> [@jqjq726](https://github.com/jqjq726) You can modify the Python script and set the labels based on video names, for example, normal events as 0, Abuse as 1, Arrest as 2, and so on. I did try some multi-class classification tasks, but the results were not very good. I feel that this might be because UCF-Crime is not a high-quality dataset. Additionally, I remember that VadCLIP involves multi-class classification tasks, so you can take a look and learn how it is implemented.

sure, thanks for your suggestion!
